### PR TITLE
Move config local to every controller in KCM

### PR DESCRIPTION
**Pull request by stewart-yu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
move controller's config from `pkg/controller/apis/config` to `pkg/controller/<controller-name>/config`,
and in that's way, every controller's type will be its own deserializable Kinds.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/daemon/config/OWNERS
Comment (sttts) [inline pkg/controller/daemon/config/OWNERS]: is this the same list as before? Can we identify the main developers of each controller via `git blame`?

File: pkg/controller/daemon/config/v1alpha1/doc.go
Comment (luxas) [inline pkg/controller/daemon/config/v1alpha1/doc.go]: shouldn't we specify here where the external types are so that the conversions are correctly generated?
right now this doesn't generate any conversions...

File: pkg/controller/garbagecollector/config/types.go
Comment (luxas) [inline pkg/controller/garbagecollector/config/types.go]: 2019 now ;)

File: pkg/controller/daemon/config/v1alpha1/register.go
Comment (luxas) [inline pkg/controller/daemon/config/v1alpha1/register.go]: I'm wondering if it would make sense to register the `addKnownTypes` and `addDefaultingFuncs` funcs to the schemebuilder by default so that every scheme & package becomes "independent" and can be added to an arbitrary scheme without any new things. The recommendeddefaults pattern would still apply but it'd also be called by a SetDefaults func...

**Review by luxas — COMMENTED** [review]
Thanks very much @stewart-yu!

Overall this looks ok. We have the struggle of wanting the config to be close to the controller, but still be publicly available. I think we need to move the conversions from `pkg/controller/apis/config` where they are now in this PR to local to the controller, using the `k8s:conversion-gen-external-types` tag per-controller.

We should make these configuration types decodable by marking them with the deepcopy runtime.Object & TypeMeta tags, idk if we should do that in this PR or not. Maybe you could make that change in a dedicated commit (in k8s.io/kube-controller-manager/config) so we can easily rip it out if we don't want to have it

The end goal here is that every controller config kind is encodable/decodable on its own. Hence every schemebuilder in every package should register what's needed for it to work, and the `pkg/controller/apis/config/scheme` package would just run `AddToScheme()` per-controller.

The OWNERS files are ok as-is for now, we can make those more suitable in a future PR.

@sttts does this sound ok?

File: pkg/controller/garbagecollector/config/types.go
Comment (stewart-yu) [reply pkg/controller/garbagecollector/config/types.go]: sorry, forgot it
will fix all comments tonight

File: pkg/controller/daemon/config/v1alpha1/register.go
Comment (stewart-yu) [reply pkg/controller/daemon/config/v1alpha1/register.go]: If need, will add it later
The key point is that the `--config` not implement in KCM. we can add at same time

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go
Comment (luxas) [inline cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go]: this must be `RecommendedDefaultServiceControllerConfiguration`, please fix this for all funcs

File: pkg/controller/daemon/config/v1alpha1/doc.go
Comment (stewart-yu) [reply pkg/controller/daemon/config/v1alpha1/doc.go]: fixed

**Review by luxas — APPROVED** [review]
/approve
Thanks @stewart-yu! This looks good now, just need a rebase :wink:

**Comment by stewart-yu** [comment]
Strange, seems the CI  blocked?

**Comment by luxas** [comment]
@stewart-yu yes, master (the commit you're on) was broken due to go.mod for a bit https://github.com/kubernetes/kubernetes/pull/72963
If you rebase again, it will be fixed :+1:

**Comment by stewart-yu** [comment]
For reviewer information:
in general, move controller's config from `pkg/controller/apis/config` to `pkg/controller/<controller-name>/config`, but three controller config specially

- `CSRSigningControllerConfiguration` to `k8s.io/kubernetes/pkg/controller/certificates/signer/config`
- `AttachDetachControllerConfiguration` to `k8s.io/kubernetes/pkg/controller/volume/attachdetach/config`
- 'PersistentVolumeBinderControllerConfiguration' to `k8s.io/kubernetes/pkg/controller/volume/persistentvolume/config`

**Review by luxas — APPROVED** [review]
Thanks @stewart-yu!
/lgtm
/approve

**Review by neolit123 — COMMENTED** [review]
thank you for working on this, stewart-yu  :+1:

**Comment by deads2k** [comment]
At a glance, this seems a little unusual.  We're gathering config together in k8s.io/kcm right?  Do we even need the internal types?

**Comment by liggitt** [comment]
I'm having a hard time following what this is doing. It looks like all the external types are still under k8s.io/kube-controller-manager/config/v1alpha1, and this moves the internal config structs from k8s.io/kubernetes/pkg/controller/apis/config to various k8s.io/kubernetes/... controller packages. Can you describe what this is intended to accomplish?

> @stewart-yu:
> every controller's type will be its own deserializable Kinds.

If the goal is to be able to serialize each controller's config individually, this PR does not accomplish that. The individual controller config structs don't have apiVersion/kind info (TypeMeta), so they cannot be serialized standalone.

> @deads2k:
> We're gathering config together in k8s.io/kcm right? Do we even need the internal types?

They will have value as we evolve the config and want to change defaults at config version boundaries.

**Comment by luxas** [comment]
> I'm having a hard time following what this is doing. It looks like all the external types are still under k8s.io/kube-controller-manager/config/v1alpha1, and this moves the internal config structs from k8s.io/kubernetes/pkg/controller/apis/config to various k8s.io/kubernetes/... controller packages. Can you describe what this is intended to accomplish?

@liggitt we want to make the controllers more self-contained. _Eventually_ also make them individually serializable etc., but this is just the first logical step; moving the code to where it makes sense. Yes, the external types are still in the "external" staging repo, and I think that's still ok. This change will make it fairly easy (eventually) to vendor just a controller, with its config and all, and bundle it in your own manager binary.

Does this explanation make sense? In other words, this ONLY moves the structs local to the controllers to start with. Then we can continue with the rest of the tasks later.

> If the goal is to be able to serialize each controller's config individually, this PR does not accomplish that. The individual controller config structs don't have apiVersion/kind info (TypeMeta), so they cannot be serialized standalone.

I know, but that's the second step, for a future PR.

>  They will have value as we evolve the config and want to change defaults at config version boundaries.

Exactly, as we graduate the kcm controllers config to beta and beyond.

**Comment by deads2k** [comment]
> > @deads2k:
> > We're gathering config together in k8s.io/kcm right? Do we even need the internal types?
> 
> They will have value as we evolve the config and want to change defaults at config version boundaries.

That depends on how many versions you have.  You need a hub, but it's less obvious that the hub needs to be internal

**Comment by luxas** [comment]
>  That depends on how many versions you have. You need a hub, but it's less obvious that the hub needs to be internal

Don't you think it makes sense for this hub to be close to the controller code?

**Comment by deads2k** [comment]
> Don't you think it makes sense for this hub to be close to the controller code?

I would keep a structure parallel to my serialization.  If I had everything to do over again, I would probably make my hub the preferred external serialization itself.  We didn't end up having different external and internal types.

**Comment by luxas** [comment]
> I would keep a structure parallel to my serialization. If I had everything to do over again, I would probably make my hub the preferred external serialization itself. We didn't end up having different external and internal types.

Hmm, I still think we're talking besides each other. I'm not sure we can "do all of API machinery again" as you imply here, but follow the patterns all the rest of k8s uses. If we keep doing what we do in other places for the controller-manager; this then makes sense, doesn't it?

@thockin @liggitt -- more thoughts?

**Comment by deads2k** [comment]
> Hmm, I still think we're talking besides each other. I'm not sure we can "do all of API machinery again" as you imply here,

That isn't what that statement says.  That statement says, "for these types, use a different hub the same way that CRDs use a different hub today".  It doesn't require rewriting machinery.

**Comment by sttts** [comment]
+1 from me to think about the hub idea. But if we want to go that route, we need to have at least one project that immediately uses that. E.g. kube-scheduler has alpha component config now (afaik), we could start promoting that to beta and select beta as the hub. We can structure the component-base code to support that.

**Comment by sttts** [comment]
@deads2k @liggitt are you ok with the overall change of this PR, i.e. splitting up the types per controller?

The topic of internal vs. hub is another independent refactoring. We would still have the conversions between hub and non-hub near to the controller code, so probably in the directories that @stewart-yu is creating here (we don't want to publish conversions, for the same reason we never wanted to publish internal types).

**Comment by luxas** [comment]
> That isn't what that statement says.

sorry that I misread you.

I agree that we could experiment with the idea of dropping internal types for config in order to get less duplicated code, but as @sttts also pointed out, that's independent from this code move. If we do drop internal types though, I guess we need to add some kind of meta field on the latest stable external version we choose as the hub that is not serialized, but only used internally for conversions to newer, experimental API versions. For example, let's say we have a `v1` API out there, and create `v2alpha1`, but still want `v1` as the hub.

**Comment by nikopen** [comment]
@stewart-yu @sttts @liggitt any chance this is merged in the following 2 weeks? 
Code Freeze is in effect from next Friday.

**Review by thockin — COMMENTED** [review]
/lgtm
/approve

Needs rebase (super minor)

**Comment by xmudrii** [comment]
A quick reminder that code freeze is approaching this Friday :snowflake: The PR was already approved and lgtm-ed, but due to required rebase it lost the `lgtm` label, so pinging for this to be checked :slightly_smiling_face:

**Comment by spiffxp** [comment]
/hold
v1.14 release lead here, this is a size/XXL PR that is attempting to land right before Code Freeze... are you _really sure_ this is a good idea?

**Review by cheftako — COMMENTED** [review]
/lgtm

**Comment by spiffxp** [comment]
/milestone clear
v1.14 release lead here, given the state of our CI signal, I'm removing this from the milestone. The size of this PR makes me wary at this late stage in the game.

Please discuss in #sig-release in slack if it's urgent this land as part of v1.14.  Otherwise it will land once we hit code thaw.

**Comment by stewart-yu** [comment]
i'm not sure whether can we remove the `/hold` tag
/cc @sttts @deads2k @luxas @thockin

**Comment by nikopen** [comment]
1.14 got branched so this can be merged as approved above

/hold clear

File: pkg/controller/apis/config/v1alpha1/defaults.go
Comment (liggitt) [inline pkg/controller/apis/config/v1alpha1/defaults.go]: these controller-specific defaults should just be normal defaulting functions so we don't have to remember to call them in every place they are used. opened https://github.com/kubernetes/kubernetes/issues/76027 to track changing these to normal defaulting functions.

**Pull request merged**
